refactor(list): add docs and improve list cmd

Mohammer5 opened the pull request:
See commit history

BREAKING CHANGE: This replaces the --pkg argument with a positional
argument of type array, so multiple packages can be provided

amcgee commented on docs/guides/applying_a_codemod.md:
````diff
@@ -0,0 +1,46 @@
+d2-utils-codemods list
+```
+
+The above command will print all available codemods, e. g. like this:
````
```suggestion
The above command will print all available codemods, like this:
```

amcgee commented on docs/guides/listing_available_codemods.md:
````diff
@@ -0,0 +1,91 @@
+d2-utils-codemods list
+```
+
+The above command will print all available codemods, e. g. like this:
````
```suggestion
This will print all available codemods, like this:
```

amcgee commented on docs/guides/listing_available_codemods.md:
````diff
@@ -0,0 +1,91 @@
+d2-utils-codemods list @dhis2/d2
+```
+
+The above command will print all available codemods, e. g. like this:
````
```suggestion
The above command will print all available codemods, like this:
```

amcgee commented on docs/guides/listing_available_codemods.md:
````diff
@@ -0,0 +1,91 @@
+d2-utils-codemods list --name change
+```
+
+The above command will print all available codemods, e. g. like this:
````
```suggestion
The above command will print all available codemods, like this:
```

amcgee commented on docs/guides/adding_a_codemod.md:
```diff
@@ -0,0 +1,13 @@
+1. figure out the correct repository<br />
+   If you want to provide a codemod for the `app-runtime` library, then the
+   library's repository is the correct location
+1. put the file into a `codemods` folder on the root level<br />
```
```suggestion
2. put the file into a `codemods` folder at the root level of the package<br />
```

Mohammer5 replied:
I changed the test, I kept the `.1` as markdown automatically increments it, so it's less hassle to add items somewhere inbetween

amcgee commented on docs/guides/adding_a_codemod.md:
```diff
@@ -0,0 +1,13 @@
+   library's repository is the correct location
+1. put the file into a `codemods` folder on the root level<br />
+   If the repository is a monorepo, make sure to put the folder into the root
+   folder of the specific workspace as it wouldn't be published otherwise
```
```suggestion
   folder of the specific workspace otherwise it will not be published
```

amcgee commented on docs/guides/adding_a_codemod.md:
```diff
@@ -0,0 +1,13 @@
+1. put the file into a `codemods` folder on the root level<br />
+   If the repository is a monorepo, make sure to put the folder into the root
+   folder of the specific workspace as it wouldn't be published otherwise
+
```
```suggestion
3. ensure that the `codemods` folder is published to `npm` with the package<br/>
    The `files` list in `package.json` must include `codemods`, and it should not
    be excluded by `.npmignore`

```

amcgee commented on docs/guides/adding_a_codemod.md:
```diff
@@ -0,0 +1,13 @@
+   If the repository is a monorepo, make sure to put the folder into the root
+   folder of the specific workspace as it wouldn't be published otherwise
+
+Currently we only support library's that are within the `@dhis2` namespace,
```
```suggestion
Currently we only support libraries within the `@dhis2` namespace on `npm`,
```

amcgee commented on docs/guides/adding_a_codemod.md:
```diff
@@ -0,0 +1,13 @@
+
+If you need to add a discoverable codemod, then you need to
+
+1. figure out the correct repository<br />
```
```suggestion
1. determine the correct repository<br />
```

amcgee commented on docs/guides/applying_a_codemod.md:
```diff
@@ -0,0 +1,46 @@
+# Applying a codemod
+
+There are two different ways to apply a codemod. Either by using a file or by
+specifying a package and codemod name.
```
```suggestion
There are two different ways to apply a codemod. It is possible to either specify a specific codemod file
or to specify a package and codemod name.
```

amcgee commented on docs/guides/applying_a_codemod.md:
```diff
@@ -0,0 +1,46 @@
+
+By supplying the path to the codemod as the first argument, you can apply said
+codemod to the files specified in the second argument (or to all files in the
+cwd, which is the default value for the files):
```
```suggestion
current working directory, if no files are specified).  Files are [`minimatch`](https://github.com/isaacs/minimatch) glob patterns:
```

amcgee commented on docs/guides/applying_a_codemod.md:
```diff
@@ -0,0 +1,46 @@
+
+## Using a package and name
+
+If one of the dhis2 packages, that you have installed, contains a codemod, you
+can use it by using the special identifier syntax: The package name, followed
+by a colon symbol, followed by the codemod name.
```
```suggestion
Codemods included in a locally-installed package within the `@dhis2` namespace
can be specified with a special identifier syntax: The package name, followed
by a `:`, followed by the codemod name.
```

amcgee commented on docs/guides/applying_a_codemod.md:
````diff
@@ -0,0 +1,46 @@
+* @dhis2/d2:31.1.0-change-imports.js
+```
+
+You can then use the entire specifier to define the codemod that should be
+applied:
````
```suggestion
You can then use the entire specifier to apply the codemod:
```

amcgee commented on docs/guides/listing_available_codemods.md:
````diff
@@ -0,0 +1,91 @@
+* @dhis2/d2:31.1.0-change-imports.js
+```
+
+Then `@dhis2/d2:31.1.0-change-imports.js` part can be copied and used:
````
```suggestion
Then `@dhis2/d2:31.1.0-change-imports.js` is a valid codemod specifier which can be applied like this:
```

amcgee commented on docs/guides/listing_available_codemods.md:
```diff
@@ -0,0 +1,91 @@
+
+## Listing all available codemods
+
+Listing the codemods can simply be done by using the `list` command:
```
```suggestion
The `list` command will list all available codemods:
```

amcgee commented on docs/guides/listing_available_codemods.md:
```diff
@@ -0,0 +1,91 @@
+
+## Filtering the codemods by name
+
+You can search the codemods by providing the `--name` argument. It will find
+all codemods if their name contains the passed in value at any position.
```
```suggestion
You can search the codemods by providing the `--name <searchKey>` argument. It will find
all codemods which contain the specified `searchKey` in their names.
```

amcgee reviewed (commented):
Some comments and suggestions, the docs suggestions are somewhat subjective

Mohammer5 commented on .gitignore:
```diff
@@ -59,3 +59,5 @@ typings/
 
 # next.js build output
 .next
```
Good question, I have no idea.. There's a chance that someone (including me) copied a generalized .gitignore from somewhere. I can't recollect doing so though

Mohammer5 replied:
I'll ignore this for now, there's lots of stuff in the .gitignore that we don't need to be in there

Mohammer5 commented on docs/guides/adding_a_codemod.md:
```diff
@@ -0,0 +1,13 @@
+   If the repository is a monorepo, make sure to put the folder into the root
+   folder of the specific workspace as it wouldn't be published otherwise
+
+Currently we only support library's that are within the `@dhis2` namespace,
+such as `@dhis2/cli-utils-cypress` or `@dhis2/ui-core`
```
Oh, yeah. I could make it a command option that defaults to '@dhis2'?

Mohammer5 replied:
I've changed this, I've added the following comment to the code (`src/utils/codemods/findAvailableCodemodsInNodeModules.js`):

```js
/**
 * Tries to find the following dirs that contain .js files:
 *   - ${NODE_MODULES}/{{folder}}/codemods/
 *   - ${NODE_MODULES}/{{namespace}}/{{folder}}/codemods/
 *
 * @param {Object} paths
 * @returns {String[]}
 */
```

Will update the docs now

Mohammer5 commented:
@amcgee I've addressed all our points, I also changed the commands to include all node modules that have a codemod, so d2 or d2-ui could be found as well (see `<root>/test/list-command.test.js` for better overview over how it works now)

amcgee commented on src/utils/codemods/findAvailableCodemodsInNodeModules.js:
```diff
@@ -2,32 +2,81 @@ const log = require('@dhis2/cli-helpers-engine').reporter
+ */
+const findCodemodsFolders = paths => {
+    return fs
+        .readdirSync(paths.NODE_MODULES)
```
As an FYI this will break completely when using I.e. yarn2 or pnpm since there's no `node_modules`.  It would be better to look at all dependencies and devDependencies and look for code mods in their virtual FS wherever that happens to be

amcgee replied:
We can do this in a separate change if needed

Mohammer5 replied:
I will try to get that in today or tomorrow after merging this one, not sure if I can manage though, I first have to figure out how to access the virtual FS, haven't touched yarn2 yet)

amcgee reviewed (commented):
The node_modules lookup gets a bit awkward when you might (or might not) have hoisting of transitive deps - and it also doesn't work in yarn2 with pnp.  We could put this off if needed to simplify this change, but the hoisting behavior should probably be "fixed" at some point (though probably not the end of the world if you end up with more codemods available than you wanted)

amcgee commented on src/utils/codemods/findAvailableCodemodsInNodeModules.js:
```diff
@@ -2,32 +2,81 @@ const log = require('@dhis2/cli-helpers-engine').reporter
+        .reduce((acc, cur) => {
+            const curPath = path.join(paths.NODE_MODULES, cur)
+
+            // Node module is not namespace and contains codemods
+            if (
+                isFolder(path.join(curPath, 'codemods')) &&
+                hasCodemods(path.join(curPath, 'codemods'))
+            ) {
+                return [...acc, cur]
+            }
+
+            // Node module is a namespace containing node modules that contain codemods
+            const curPathFolders = fs
+                .readdirSync(curPath)
+                .filter(
+                    name =>
+                        isNotDotFolder(name) &&
+                        isFolder(path.join(curPath, name, 'codemods')) &&
+                        hasCodemods(path.join(curPath, name, 'codemods'))
+                )
+                .map(name => `${cur}/${name}`)
```
I *think* this is actually incorrect logic... namespaces in npm must always start with @, but this would also traverse a second level of the node_modules directory tree for modules that just don't have their own codemods dir - the first condition would fail for i.e. `lodash` because there's no `codemods` directory there, but the algorithm would continue to traverse all sub-directories in the `lodash` package looking for one that contains `codemods`.  This could result in false-positives (though unlikely) and would also be quite expensive.

amcgee commented on src/utils/codemods/findAvailableCodemodsInNodeModules.js:
```diff
@@ -32,45 +49,62 @@ const hasCodemods = codemodsPath =>
-
-            return [...acc, ...curPathFolders]
-        }, [])
+        packageJson = JSON.parse(packageJsonRaw)
```
Any reason not to just use `require(paths.PACKAGE_JSON)` here?  It parses automatically

Mohammer5 replied:
Changed

amcgee commented on src/utils/codemods/findAvailableCodemodsInNodeModules.js:
```diff
@@ -32,45 +49,62 @@ const hasCodemods = codemodsPath =>
+
+        const codemodPath = path.join(depPath, 'codemods')
+
+        if (!isFolder(codemodPath)) {
+            return acc
+        }
+
+        if (!hasCodemods(codemodPath)) {
+            return acc
+        }
```
*minor* could be just one `if` clause

amcgee commented on src/utils/codemods/findAvailableCodemodsInNodeModules.js:
```diff
@@ -32,45 +49,62 @@ const hasCodemods = codemodsPath =>
+            return acc
+        }
+
+        return [...acc, dependency]
```
*minor* it's sightly more performant to mutate the result:

```
acc.push(dependency)
return acc
```

But that's less "functional", shouldn't be a big deal either way here

Mohammer5 replied:
Shouldn't matter in this case

amcgee commented on src/utils/makePaths.js:
```diff
@@ -1,6 +1,6 @@
 module.exports.makePaths = cwd => {
-    const NODE_MODULES = path.join(cwd, 'node_modules')
-    return { NODE_MODULES }
+    const PACKAGE_JSON = path.join(cwd, 'package.json')
```
It might be nice if we used `find-up` to run the script from the nearest `package.json`, but that might also be confusing when using file globs relative to the actual `cwd`?

Mohammer5 replied:
I like the idea, sounds more like a future PR though

amcgee commented on src/utils/codemods/findAvailableCodemodsInNodeModules.js:
```diff
@@ -2,32 +2,115 @@ const log = require('@dhis2/cli-helpers-engine').reporter
+    try {
+        const resolvedPath = require.resolve(
+            packageJsonResolvePath,
+            process.env.NODE_ENV === 'test'
+                ? { paths: [process.env.RESOLVE_PATH] }
+                : undefined
```
Should this always be `paths.cwd`?  That way it resolves relative to wherever the command is run (even if cwd is passed on the command-line) and we don't need a special case for tests

Mohammer5 replied:
In normal use cases it needs to be `${paths.cwd}/node_modules`, when testing, it needs to be `${paths.cwd}/test/node_modules` for the integration tests and  `src/utils/codemods/__test__/findAvailableCodemodsInNodeModules/` for a unit test

Mohammer5 replied:
I did it this way because you can't mock `require.resolve` with jest

Mohammer5 replied:
ping @amcgee

amcgee replied:
@Mohammer5 - it's common practice in our cli modules to support a `--cwd path/to/cwd` argument - would be nice to support this (falling back to `process.cwd()`) and pass that to this function, then you can just pass a `cwd` argument to this function in jest and don't need any test-specific code.

Mohammer5 replied:
In this case this has nothing to do with the "normal" use case, it's for testing only.

The normal use-case is: The `findAvailableCodemodsInNodeModules` function uses the `--cwd` option (defaults to `process.cwd`) (see [here](https://github.com/dhis2/cli-utils-codemods/blob/6374f2d53903188323aef4113121b93cb3b36e0b/src/commands/list.js#L23)) to locate the `package.json`.
The function then extracts the dependencies from the package.json so it can iterate through the found packages to find only those that contain codemods. In order to find the correct location, it uses `require.resolve`.

When testing this function with jest, I need to have some modules with codemods in the `node_modules` folder, which is not guaranteed unless when providing it in a mocked node_modules folder, hence the reason to mock `require.resolve`.

You can't mock `require.resolve`, which is the reason for the check

Mohammer5 replied:
@amcgee and I had a quick chat on slack and the issue was that `require.resolve` won't respect the `--cwd` option when passed in.

This has been addressed and there's no different value for test environments anymore. Instad the cwd is passed as a path to the `resolve` method.

amcgee commented on README.md:
```diff
@@ -0,0 +1,19 @@
+
+Deep links:
+
+-   [Bug](https://jira.dhis2.org/secure/CreateIssueDetails!init.jspa?pid=10703&issuetype=10006&components=11016)
+-   [Feature](https://jira.dhis2.org/secure/CreateIssueDetails!init.jspa?pid=10703&issuetype=10300&components=11016)
+-   [Task](https://jira.dhis2.org/secure/CreateIssueDetails!init.jspa?pid=10703&issuetype=10003&components=11016)
```
Can be fixed later, but these links are for the `utils-cypress` component

amcgee replied:
```suggestion
-   [Bug](https://jira.dhis2.org/secure/CreateIssueDetails!init.jspa?pid=10703&issuetype=10006&components=11031)
-   [Feature](https://jira.dhis2.org/secure/CreateIssueDetails!init.jspa?pid=10703&issuetype=10300&components=11031)
-   [Task](https://jira.dhis2.org/secure/CreateIssueDetails!init.jspa?pid=10703&issuetype=10003&components=11031)
```

amcgee replied:
Updated the links ^

amcgee commented on src/utils/codemods/findAvailableCodemodsInNodeModules.js:
```diff
@@ -2,32 +2,103 @@ const log = require('@dhis2/cli-helpers-engine').reporter
+        const resolvedPath = require.resolve(packageJsonResolvePath, { paths })
+        return path.dirname(resolvedPath)
+    } catch (e) {
+        return ''
```
Any reason this is an empty string instead of `null` or `undefined`?

Mohammer5 replied:
Yeah. It allows to avoid null checks. In really doesn't matter in this case, it's just a habit to never change the type of a (return) value as that could introduce unforeseen type bugs in my experience

amcgee commented on src/utils/codemods/findAvailableCodemodsInNodeModules.js:
```diff
@@ -2,32 +2,103 @@ const log = require('@dhis2/cli-helpers-engine').reporter
+        const curPath = path.join(codemodsPath, curFile)
+
+        if (isFolder(curPath)) return false
+        if (!curFile.match(/\.(j|t)sx?$/)) return false
```
(minor) this should probably use the `isDotJSFile` helper (and update the regex) right?

amcgee replied:
Yeah, looks like it should - a `.tsx` file will pass here but fail later.  Also not sure what a `.jsx` or `.tsx` file would mean here.... also also not sure we support typescript?

Mohammer5 replied:
Ah, no idea how that happened. Updated

amcgee commented on test/apply-command.test.js:
```diff
@@ -6,6 +6,32 @@ const spawn = require('cross-spawn')
+    }
+
+    if (result.stderr && result.stderr.toString()) {
+        result.strout && console.log('STDOUT:')
+        result.strout && console.log(result.stdout.toString())
+        console.log('STDERR:')
+        console.log(result.stderr.toString())
```
Think I mentioned this before... is there a reason not to use `exec` from `@dhis2/cli-helpers-engine` here?  It already supports piping stdout and stderr to the parent (this dumps stderr to stdout which seems incorrect?). If we want to use `cross-spawn` we should update the helper library instead

amcgee replied:
Ah sorry I see this is a test.... nevermind - maybe we should snapshot stdout and stderr rather than just dumping it to console.log though?

Mohammer5 replied:
See comment below

amcgee commented on test/list-command.test.js:
```diff
@@ -1,28 +1,101 @@
+    )
+
+    if (result.stderr.toString()) {
+        console.log('STDOUT:')
+        console.log(result.stdout.toString())
+        console.log('STDERR:')
+        console.log(result.stderr.toString())
```
Probably this should also be a `snapshot` or `inlineSnapshot`?

amcgee replied:
I guess we're actuually doing that later, but it seems incorrect too console.log stderr here?

Mohammer5 replied:
Love the idea! Updated

amcgee reviewed (changes requested):
Sorry for the nitpicking, few more comments!

amcgee reviewed (approved):
LGTM 🚀
